shadowsocks/tcp.go

```diff
@@ -165,14 +165,15 @@ func proxyConnection(clientConn onet.DuplexConn, proxyMetrics *metrics.ProxyMetr
 	return nil
 }
 
-func (s *tcpService) Start() {
+func (s *tcpService) Start() error {
+	errorChan := make(chan *onet.ConnectionError)
```

Comment:
I'm not comfortable with the introduction of this error channel.
It's exposing something that is too low level and the semantics is confusing. Start shouldn't return an error on success, even if there are probes. And I remember having issues with not draining channels correctly.

The most important thing to test is the behavior. You can do that by treating the service as a black box.

If you want to test that the metrics are set correctly, pass a fake metrics object to the service, and check the probe metrics in the end of the test.

Reply:
That's fair. I am not sure how to proceed though - the issue is that I don't see how to directly test whether a timeout has occurred while treating the service as a black box, i.e., after I call s.Start(), how can I observe whether the session is still open, since the only functions I have available are Start and Stop?

Reply:
Overhauled test so it now checks how long a read on a connection blocks. This should be much less invasive!